File: php/app/core/Ai.php
<?php
/**
 * Ai.php - Trợ lý AI tương thích OpenAI (dùng cho gợi ý phân luồng tay)
 * Hệ thống phân luồng Mail công vụ - Thiết kế bởi Trương Anh Tuấn
 */

class Ai
{
    const MAX_TOKENS_TRAN = 64000;
    const TIMEOUT_TRAN    = 300;

    public static function bat(): bool
    {
        return Ung::bat('ai.bat', false)
            && trim((string)Ung::get('ai.url')) !== ''
            && trim((string)Ung::get('ai.model')) !== '';
    }

    public static function urlChat(?string $goc = null): string
    {
        $u = rtrim(trim((string)($goc ?? Ung::get('ai.url'))), '/');
        if ($u === '') return '';
        if (strpos($u, '/chat/completions') !== false) return $u;
        if (substr($u, -12) === '/completions') return $u;
        return $u . '/chat/completions';
    }

    public static function maxTokens(): int
    {
        $n = (int)Ung::so('ai.max_tokens', 4096);
        return max(1, min(self::MAX_TOKENS_TRAN, $n));
    }

    public static function timeout(): int
    {
        $n = (int)Ung::so('ai.timeout', 60);
        return max(1, min(self::TIMEOUT_TRAN, $n));
    }

    /**
     * Gọi API chat. Trả về [thành công, nội dung|thông báo lỗi, thông tin thêm]
     */
    public static function chat(array $tinNhan, array $ghiDe = []): array
    {
        $url = self::urlChat($ghiDe['url'] ?? null);
        if ($url === '') return [false, 'Chưa khai báo địa chỉ dịch vụ AI', []];

        $than = [
            'model'       => (string)($ghiDe['model'] ?? Ung::get('ai.model')),
            'messages'    => $tinNhan,
            'temperature' => (float)($ghiDe['temperature'] ?? Ung::get('ai.temperature', '0.1')),
            'max_tokens'  => (int)($ghiDe['max_tokens'] ?? self::maxTokens()),
            'stream'      => false,
        ];
        $khoa = (string)($ghiDe['api_key'] ?? Ung::get('ai.api_key'));
        $timeout = (int)($ghiDe['timeout'] ?? self::timeout());
        $timeout = max(1, min(self::TIMEOUT_TRAN, $timeout));

        $tieuDe = ['Content-Type: application/json; charset=utf-8', 'Accept: application/json'];
        if ($khoa !== '') $tieuDe[] = 'Authorization: Bearer ' . $khoa;

        if (!function_exists('curl_init')) {
            return [false, 'Máy chủ web chưa bật phần mở rộng cURL của PHP', []];
        }

        $t0 = microtime(true);
        $ch = curl_init($url);
        curl_setopt_array($ch, [
            CURLOPT_POST           => true,
            CURLOPT_POSTFIELDS     => json_encode($than, JSON_UNESCAPED_UNICODE),
            CURLOPT_HTTPHEADER     => $tieuDe,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => $timeout,
            CURLOPT_CONNECTTIMEOUT => min(30, $timeout),
            CURLOPT_FOLLOWLOCATION => true,
            CURLOPT_USERAGENT      => 'PhanLuongMail/1.0',
        ]);
        $traLoi = curl_exec($ch);
        $maHttp = (int)curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
        $loiCurl = curl_error($ch);
        curl_close($ch);
        $giay = round(microtime(true) - $t0, 2);

        if ($traLoi === false) return [false, 'Lỗi kết nối AI: ' . $loiCurl, ['giay' => $giay]];

        $j = json_decode($traLoi, true);
        if (!is_array($j)) {
            return [false, 'Dịch vụ AI trả về dữ liệu không hợp lệ: ' . mb_substr((string)$traLoi, 0, 300),
                    ['giay' => $giay]];
        }
        if (isset($j['error'])) {
            $m = is_array($j['error']) ? ($j['error']['message'] ?? json_encode($j['error'])) : $j['error'];
            return [false, 'AI báo lỗi: ' . $m, ['giay' => $giay]];
        }
        if ($maHttp < 200 || $maHttp >= 300) {
            return [false, "Dịch vụ AI trả về HTTP {$maHttp}: " . mb_substr((string)$traLoi, 0, 300),
                    ['giay' => $giay]];
        }
        $nd = $j['choices'][0]['message']['content'] ?? ($j['choices'][0]['text'] ?? '');
        if ($nd === '') return [false, 'AI trả về nội dung rỗng', ['giay' => $giay]];

        return [true, (string)$nd, ['giay' => $giay, 'token' => (int)($j['usage']['total_tokens'] ?? 0)]];
    }

    /** Kiểm tra kết nối */
    public static function kiemTra(array $ghiDe = []): array
    {
        [$ok, $nd, $tt] = self::chat(
            [['role' => 'user', 'content' => 'Trả lời đúng một từ: OK']],
            array_merge($ghiDe, ['max_tokens' => 16, 'temperature' => 0])
        );
        if (!$ok) return [false, $nd];
        return [true, sprintf('Kết nối AI thành công (%.1f giây). Mô hình trả lời: "%s"',
                              $tt['giay'] ?? 0, mb_substr(trim($nd), 0, 60))];
    }

    /**
     * Nhờ AI đề xuất mã cho một email (dùng ở màn hình phân luồng tay).
     * @return array [thành công, dữ liệu|thông báo lỗi]
     */
    public static function goiYPhanLuong(array $email, array $tenTep = []): array
    {
        if (!self::bat()) return [false, 'Chức năng AI đang tắt. Bật trong mục Cài đặt.'];

        $dm = LuuMail::danhMuc();
        $p = "DANH MỤC TRƯỜNG (mã | tên):\n";
        foreach (array_slice($dm['truong'], 0, 400) as $t) $p .= $t['ma'] . ' | ' . $t['ten'] . "\n";
        $p .= "\nDANH MỤC NGƯỜI XỬ LÝ (mã | họ tên):\n";
        foreach (array_slice($dm['nguoi_xu_ly'], 0, 200) as $t) $p .= $t['ma'] . ' | ' . $t['ho_ten'] . "\n";
        $p .= "\nDANH MỤC MÃ VĂN BẢN (mã | tên) - có thể chưa đầy đủ:\n";
        foreach (array_slice($dm['van_ban'], 0, 300) as $t) $p .= $t['ma'] . ' | ' . $t['ten'] . "\n";

        $p .= "\n===== THÔNG TIN EMAIL CẦN PHÂN LUỒNG =====\n";
        $p .= 'Người gửi: ' . ($email['ten_nguoi_gui'] ?? '') . ' <' . ($email['nguoi_gui'] ?? '') . ">\n";
        $p .= 'Tiêu đề: ' . ($email['tieu_de'] ?? '') . "\n";
        $p .= 'Thời gian: ' . Util::ngay($email['ngay_gui'] ?? null) . " (giờ Việt Nam)\n";
        $p .= "Tệp đính kèm:\n";
        if (!$tenTep) $p .= "  (không có)\n";
        foreach ($tenTep as $t) $p .= '  - ' . $t . "\n";
        $p .= "Trích nội dung:\n" . mb_substr(trim((string)($email['noi_dung_text'] ?: ($email['doan_trich'] ?? ''))), 0, 3000) . "\n";

        $p .= "\n===== YÊU CẦU =====\n"
            . "Quy ước đặt tên: <mã trường>_<mã văn bản>_<mã người xử lý>, ví dụ 001_001_TAT.\n"
            . "Hãy xác định 3 mã đó. Quy tắc:\n"
            . "1. Ưu tiên mã đọc được từ tên tệp, sau đó tới tiêu đề, cuối cùng mới suy luận từ nội dung.\n"
            . "2. ma_truong và ma_nguoi_xu_ly BẮT BUỘC nằm trong danh mục. Không chắc thì để chuỗi rỗng.\n"
            . "3. ma_van_ban có thể là mã mới - cứ ghi đúng mã đọc/suy được.\n"
            . "4. do_tin_cay là số thực 0..1.\n"
            . "CHỈ TRẢ LỜI bằng một đối tượng JSON duy nhất theo khuôn:\n"
            . '{"ma_truong":"","ma_van_ban":"","ma_nguoi_xu_ly":"","do_tin_cay":0.0,"ly_do":"giải thích ngắn bằng tiếng Việt"}';

        [$ok, $nd, $tt] = self::chat([
            ['role' => 'system', 'content' => (string)Ung::get('ai.nhac_he_thong',
                'Bạn là trợ lý văn thư của Sở Giáo dục và Đào tạo. Chỉ trả lời bằng JSON.')],
            ['role' => 'user', 'content' => $p],
        ]);
        if (!$ok) return [false, $nd];

        $json = self::bocJson($nd);
        $kq = json_decode($json, true);
        if (!is_array($kq)) return [false, 'Không đọc được JSON trong câu trả lời của AI: ' . mb_substr($nd, 0, 300)];

        $tin = (float)($kq['do_tin_cay'] ?? 0);
        if ($tin > 1) $tin /= 100;
        return [true, [
            'ma_truong'      => trim((string)($kq['ma_truong'] ?? '')),
            'ma_van_ban'     => trim((string)($kq['ma_van_ban'] ?? '')),
            'ma_nguoi_xu_ly' => strtoupper(trim((string)($kq['ma_nguoi_xu_ly'] ?? ''))),
            'do_tin_cay'     => max(0, min(1, $tin)),
            'ly_do'          => mb_substr(trim((string)($kq['ly_do'] ?? '')), 0, 2000),
            'giay'           => $tt['giay'] ?? 0,
        ]];
    }

    private static function bocJson(string $s): string
    {
        $t = trim($s);
        if (($f = strpos($t, '```')) !== false) {
            $bd = strpos($t, "\n", $f);
            $kt = strrpos($t, '```');
            if ($bd !== false && $kt !== false && $kt > $bd) $t = trim(substr($t, $bd + 1, $kt - $bd - 1));
        }
        $a = strpos($t, '{');
        $b = strrpos($t, '}');
        if ($a !== false && $b !== false && $b > $a) return substr($t, $a, $b - $a + 1);
        return $t;
    }
}

File: cpp/src/ai.cpp
// =====================================================================
//  ai.cpp - Trợ lý AI tương thích OpenAI
// =====================================================================
#include "ai.h"
#include "http_client.h"
#include "json.h"
#include "util.h"
#include "nhatky.h"

#include <algorithm>
#include <cstdio>

namespace mr {

void CauHinhAi::chuanHoa() {
    url = trim(url);
    while (!url.empty() && url.back() == '/') url.pop_back();
    model = trim(model);
    api_key = trim(api_key);
    if (max_tokens <= 0) max_tokens = 4096;
    if (max_tokens > MAX_TOKENS_TRAN) max_tokens = MAX_TOKENS_TRAN;
    if (timeout <= 0) timeout = 60;
    if (timeout > TIMEOUT_TRAN) timeout = TIMEOUT_TRAN;
    if (temperature < 0) temperature = 0;
    if (temperature > 2) temperature = 2;
    if (nguong_tin_cay < 0) nguong_tin_cay = 0;
    if (nguong_tin_cay > 1) nguong_tin_cay = 1;
    if (nhac_he_thong.empty())
        nhac_he_thong = "Bạn là trợ lý văn thư của Sở Giáo dục và Đào tạo. Nhiệm vụ: đọc thông tin "
                        "email và xác định MÃ TRƯỜNG, MÃ VĂN BẢN, MÃ NGƯỜI XỬ LÝ dựa trên danh mục "
                        "được cung cấp. Chỉ trả lời bằng JSON.";
}

CauHinhAi CauHinhAi::tuCauHinh(const CauHinh& ch) {
    CauHinhAi c;
    c.bat            = toBool(ch.uuTien("ai.bat", "ai.bat", "0"), false);
    c.url            = ch.uuTien("ai.url", "ai.url", "https://api.openai.com/v1");
    c.api_key        = ch.uuTien("ai.api_key", "ai.api_key", "");
    c.model          = ch.uuTien("ai.model", "ai.model", "gpt-4o-mini");
    c.max_tokens     = (int)toLL(ch.uuTien("ai.max_tokens", "ai.max_tokens", "4096"), 4096);
    c.timeout        = (int)toLL(ch.uuTien("ai.timeout", "ai.timeout", "60"), 60);
    c.temperature    = toDouble(ch.uuTien("ai.temperature", "ai.temperature", "0.1"), 0.1);
    c.nguong_tin_cay = toDouble(ch.uuTien("ai.nguong_tin_cay", "ai.nguong_tin_cay", "0.6"), 0.6);
    c.nhac_he_thong  = ch.uuTien("ai.nhac_he_thong", "ai.nhac_he_thong", "");
    c.chuanHoa();
    return c;
}

std::string TroLyAi::urlChat(const std::string& gocIn) {
    std::string goc = trim(gocIn);
    while (!goc.empty() && goc.back() == '/') goc.pop_back();
    if (goc.empty()) return "";
    if (goc.find("/chat/completions") != std::string::npos) return goc;
    if (endsWith(goc, "/completions")) return goc;
    return goc + "/chat/completions";
}

// ---------------------------------------------------------------------
//  Dựng câu nhắc
// ---------------------------------------------------------------------
std::string TroLyAi::dungPrompt(const BanGhiEmail& em, const DanhMuc& dm) const {
    std::string p;
    p += "DANH MỤC TRƯỜNG (mã | tên):\n";
    size_t n = 0;
    for (const auto& t : dm.truong) {
        if (n++ >= 400) { p += "... (còn nữa)\n"; break; }
        p += t.ma + " | " + t.ten + "\n";
    }
    p += "\nDANH MỤC NGƯỜI XỬ LÝ (mã | họ tên):\n";
    n = 0;
    for (const auto& x : dm.nguoi) {
        if (n++ >= 200) { p += "... (còn nữa)\n"; break; }
        p += x.ma + " | " + x.ho_ten + "\n";
    }
    p += "\nDANH MỤC MÃ VĂN BẢN (mã | tên) - danh mục này KHÔNG đầy đủ, "
         "nếu email dùng mã khác thì vẫn lấy đúng mã đọc được:\n";
    n = 0;
    for (const auto& v : dm.van_ban) {
        if (n++ >= 300) { p += "... (còn nữa)\n"; break; }
        p += v.ma + " | " + v.ten + "\n";
    }

    p += "\n===== THÔNG TIN EMAIL CẦN PHÂN LUỒNG =====\n";
    p += "Người gửi: " + em.ten_nguoi_gui + " <" + em.nguoi_gui + ">\n";
    p += "Tiêu đề: " + em.tieu_de + "\n";
    p += "Thời gian: " + dinhDangGioVN(em.ngay_gui) + " (giờ Việt Nam)\n";
    p += "Tệp đính kèm:\n";
    if (em.tep.empty()) p += "  (không có)\n";
    for (const auto& t : em.tep) p += "  - " + t.ten + " (" + dinhDangDungLuong(t.dung_luong) + ")\n";
    p += "Trích nội dung:\n" + catUtf8(trim(em.noi_dung_text.empty() ? em.doan_trich : em.noi_dung_text), 3000) + "\n";

    p += "\n===== YÊU CẦU =====\n"
         "Quy ước đặt tên: <mã trường>_<mã văn bản>_<mã người xử lý>, ví dụ 001_001_TAT.\n"
         "Hãy xác định 3 mã đó cho email trên. Quy tắc:\n"
         "1. Ưu tiên mã đọc được trực tiếp từ tên tệp đính kèm, sau đó tới tiêu đề, cuối cùng mới suy luận từ nội dung.\n"
         "2. ma_truong và ma_nguoi_xu_ly BẮT BUỘC phải nằm trong danh mục ở trên. Nếu không chắc chắn, để chuỗi rỗng.\n"
         "3. ma_van_ban có thể là mã mới không có trong danh mục - cứ ghi đúng mã đọc/suy được.\n"
         "4. do_tin_cay là số thực 0..1 thể hiện mức độ chắc chắn của bạn.\n"
         "CHỈ TRẢ LỜI bằng một đối tượng JSON duy nhất, không kèm giải thích ngoài JSON, theo đúng khuôn:\n"
         "{\"ma_truong\":\"\",\"ma_van_ban\":\"\",\"ma_nguoi_xu_ly\":\"\",\"do_tin_cay\":0.0,\"ly_do\":\"giải thích ngắn bằng tiếng Việt\"}";
    return p;
}

// ---------------------------------------------------------------------
//  Bóc JSON khỏi câu trả lời của mô hình
// ---------------------------------------------------------------------
static std::string bocJson(const std::string& s) {
    std::string t = trim(s);
    // Bỏ rào ```json ... ```
    size_t f = t.find("```");
    if (f != std::string::npos) {
        size_t bd = t.find('\n', f);
        size_t kt = t.rfind("```");
        if (bd != std::string::npos && kt != std::string::npos && kt > bd)
            t = trim(t.substr(bd + 1, kt - bd - 1));
    }
    size_t a = t.find('{');
    size_t b = t.rfind('}');
    if (a != std::string::npos && b != std::string::npos && b > a) return t.substr(a, b - a + 1);
    return t;
}

// ---------------------------------------------------------------------
//  Gọi AI
// ---------------------------------------------------------------------
bool TroLyAi::doanMa(const BanGhiEmail& em, const DanhMuc& dm, KetQuaAi& ra, std::string& loi) {
    ra = KetQuaAi();
    if (!bat()) { loi = "Chức năng AI đang tắt"; return false; }

    std::string url = urlChat(ch_.url);
    if (url.empty()) { loi = "Chưa khai báo địa chỉ dịch vụ AI"; return false; }

    Json tinNhan = Json::mang();
    Json m1 = Json::doiTuong();
    m1.dat("role", "system");
    m1.dat("content", ch_.nhac_he_thong);
    tinNhan.them(m1);
    Json m2 = Json::doiTuong();
    m2.dat("role", "user");
    m2.dat("content", dungPrompt(em, dm));
    tinNhan.them(m2);

    Json than = Json::doiTuong();
    than.dat("model", ch_.model);
    than.dat("messages", tinNhan);
    than.dat("temperature", ch_.temperature);
    than.dat("max_tokens", (long long)ch_.max_tokens);
    than.dat("stream", false);

    std::vector<std::string> hd;
    if (!ch_.api_key.empty()) hd.push_back("Authorization: Bearer " + ch_.api_key);
    hd.push_back("Accept: application/json");

    int64_t t0 = nowEpochMs();
    HttpPhanHoi pr = HttpClient::postJson(url, than.ketXuat(), hd, ch_.timeout);
    ra.giay = (nowEpochMs() - t0) / 1000.0;

    if (!pr.loi.empty()) { loi = pr.loi; return false; }
    if (pr.than.empty()) { loi = "Dịch vụ AI không trả về dữ liệu (HTTP " + std::to_string(pr.ma) + ")"; return false; }

    Json j;
    if (!Json::phanTich(pr.than, j, nullptr)) {
        loi = "Dịch vụ AI trả về dữ liệu không phải JSON: " + catUtf8(pr.than, 300);
        return false;
    }
    if (j.coKhoa("error")) {
        const Json& e = j.lay("error");
        loi = "AI báo lỗi: " + e.lay("message").chuoi(catUtf8(pr.than, 200));
        return false;
    }
    if (pr.ma < 200 || pr.ma >= 300) {
        loi = "Dịch vụ AI trả về HTTP " + std::to_string(pr.ma) + ": " + catUtf8(pr.than, 300);
        return false;
    }

    std::string noiDung = j.duongDan("choices.0.message.content").chuoi();
    if (noiDung.empty()) noiDung = j.duongDan("choices.0.text").chuoi();
    if (noiDung.empty()) {
        loi = "AI trả về nội dung rỗng";
        return false;
    }
    ra.tra_loi_tho = catUtf8(noiDung, 4000);
    ra.token_da_dung = (int)j.duongDan("usage.total_tokens").nguyen(0);

    Json kq;
    if (!Json::phanTich(bocJson(noiDung), kq, nullptr) || !kq.laDoiTuong()) {
        loi = "Không đọc được JSON trong câu trả lời của AI: " + catUtf8(noiDung, 300);
        return false;
    }

    ra.ma.truong  = trim(kq.lay("ma_truong").chuoi());
    ra.ma.van_ban = trim(kq.lay("ma_van_ban").chuoi());
    ra.ma.nguoi   = toUpper(trim(kq.lay("ma_nguoi_xu_ly").chuoi()));
    ra.do_tin_cay = kq.lay("do_tin_cay").so(0);
    if (ra.do_tin_cay > 1.0) ra.do_tin_cay = ra.do_tin_cay / 100.0;   // mô hình trả 0..100
    if (ra.do_tin_cay < 0) ra.do_tin_cay = 0;
    if (ra.do_tin_cay > 1) ra.do_tin_cay = 1;
    ra.giai_thich = catUtf8(trim(kq.lay("ly_do").chuoi()), 2000);
    return true;
}

bool TroLyAi::kiemTra(std::string& thongDiep, std::string& loi) {
    std::string url = urlChat(ch_.url);
    if (url.empty()) { loi = "Chưa khai báo địa chỉ dịch vụ AI"; return false; }

    Json tinNhan = Json::mang();
    Json m = Json::doiTuong();
    m.dat("role", "user");
    m.dat("content", "Trả lời đúng một từ: OK");
    tinNhan.them(m);

    Json than = Json::doiTuong();
    than.dat("model", ch_.model);
    than.dat("messages", tinNhan);
    than.dat("temperature", 0);
    than.dat("max_tokens", (long long)16);

    std::vector<std::string> hd;
    if (!ch_.api_key.empty()) hd.push_back("Authorization: Bearer " + ch_.api_key);

    HttpPhanHoi pr = HttpClient::postJson(url, than.ketXuat(), hd,
                                          std::min(ch_.timeout, 60));
    if (!pr.loi.empty()) { loi = pr.loi; return false; }

    Json j;
    Json::phanTich(pr.than, j, nullptr);
    if (j.coKhoa("error")) {
        loi = "AI báo lỗi: " + j.duongDan("error.message").chuoi(catUtf8(pr.than, 200));
        return false;
    }
    if (pr.ma < 200 || pr.ma >= 300) {
        loi = "Dịch vụ AI trả về HTTP " + std::to_string(pr.ma) + ": " + catUtf8(pr.than, 300);
        return false;
    }
    std::string nd = trim(j.duongDan("choices.0.message.content").chuoi());
    char buf[256];
    std::snprintf(buf, sizeof(buf), "Kết nối AI thành công (%.1f giây). Mô hình trả lời: \"%s\"",
                  pr.giay, catUtf8(nd, 60).c_str());
    thongDiep = buf;
    return true;
}

} // namespace mr
